api/census.go

```diff
@@ -0,0 +1,201 @@
+	invalid, err := ref.InsertBatch(keys, values)
+	if err != nil {
+		ErrGenericInternalServerError.WithErr(err).Write(w)
+		return
+	}
+	if len(invalid) > 0 {
+		ErrMalformedBody.WithErr(fmt.Errorf("failed to insert %d participants", len(invalid))).Write(w)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (a *API) getCensusParticipants(w http.ResponseWriter, r *http.Request) {
+	censusID, err := uuid.Parse(r.URL.Query().Get("id"))
+	if err != nil {
+		ErrMalformedBody.WithErr(err).Write(w)
+		return
+	}
+
+	ref, err := a.storage.CensusDB().Load(censusID)
+	if err != nil {
+		ErrGenericInternalServerError.WithErr(err).Write(w)
+		return
+	}
+
+	// TODO: Implement pagination properly
+	keys := make([][]byte, 0)
+	values := make([][]byte, 0)
+	err = ref.Tree().Iterate(nil, func(k []byte, v []byte) {
+		keys = append(keys, k)
+		values = append(values, v)
+	})
+	if err != nil {
+		ErrGenericInternalServerError.WithErr(err).Write(w)
+		return
+	}
+
+	participants := make([]CensusParticipant, len(keys))
+	for i := range keys {
+		participants[i] = CensusParticipant{
+			Key:    keys[i],
+			Weight: (*types.BigInt)(arbo.BytesToBigInt(values[i])),
+		}
+	}
+
+	httpWriteJSON(w, map[string]interface{}{
+		"participants": participants,
+	})
+}
+
+func (a *API) getCensusRoot(w http.ResponseWriter, r *http.Request) {
+	censusID, err := uuid.Parse(r.URL.Query().Get("id"))
+	if err != nil {
+		ErrInvalidCensusID.WithErr(err).Write(w)
+		return
+	}
+
+	ref, err := a.storage.CensusDB().Load(censusID)
+	if err != nil {
+		ErrGenericInternalServerError.WithErr(err).Write(w)
+		return
+	}
+
+	httpWriteJSON(w, map[string]types.HexBytes{
+		"root": types.HexBytes(ref.Root()),
+	})
+}
+
+func (a *API) getCensusSize(w http.ResponseWriter, r *http.Request) {
+	censusID, err := uuid.Parse(r.URL.Query().Get("id"))
+	if err != nil {
+		ErrInvalidCensusID.WithErr(err).Write(w)
+		return
+	}
+
+	ref, err := a.storage.CensusDB().Load(censusID)
+	if err != nil {
+		ErrGenericInternalServerError.WithErr(err).Write(w)
+		return
+	}
+
+	httpWriteJSON(w, map[string]interface{}{
+		"size": ref.Size(),
+	})
+}
+
+func (a *API) deleteCensus(w http.ResponseWriter, r *http.Request) {
+	censusID, err := uuid.Parse(r.URL.Query().Get("id"))
+	if err != nil {
+		ErrInvalidCensusID.WithErr(err).Write(w)
+		return
+	}
+
+	if err := a.storage.CensusDB().Del(censusID); err != nil {
+		ErrGenericInternalServerError.WithErr(err).Write(w)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (a *API) getCensusProof(w http.ResponseWriter, r *http.Request) {
+	rootHex := r.URL.Query().Get("root")
+	root, err := hex.DecodeString(rootHex)
+	if err != nil {
+		ErrInvalidCensusID.WithErr(err).Write(w)
+		return
+	}
+
+	keyHex := r.URL.Query().Get("key")
+	key, err := hex.DecodeString(keyHex)
+	if err != nil {
+		ErrMalformedBody.WithErr(err).Write(w)
+		return
+	}
+
+	leafKey := key
+	if len(key) > census.KeyMaxLen {
+		leafKey = a.storage.CensusDB().HashAndTrunkKey(key)
```

Comment:
This must be reviewed @lucasmenendez @altergui 

If we use 160 levels, we must trunk at 160/8 = 20 the keys (addresses). 

So the merkle-tree indices are `mimcBN254(address)[:20]`. Are we in the same page?

Reply:
Or, if we use Addresses which are already 20 bytes long, we might be fine. Then its not necessary to make the hash?

Reply:
Ok. I've verified it works fine using 20 bytes addresses. So the Hash and Trunk is only used if the key size > 20 bytes.